aderyn_core/src/detect/low/unused_imports.rs

```diff
@@ -0,0 +1,297 @@
+                        imported_symbols,
+                    );
+                }
+            }
+        }
+
+        for source_unit in context.source_units() {
+            let referenced_declarations = ExtractReferencedDeclarationsConditionally::from(
+                source_unit,
+                context,
+                Box::new(|node_id, context| {
+                    if let Some(node) = context.nodes.get(&node_id) {
+                        return node
+                            .closest_ancestor_of_type(context, NodeType::ImportDirective)
+                            .is_none();
+                    }
+                    false
+                }),
+            )
+            .extracted;
+
+            for referenced_symbol in referenced_declarations {
+                graph.mark_used_pathways(source_unit.id, referenced_symbol);
+                if let Some(symbol_place) = context.nodes.get(&referenced_symbol) {
+                    if let Some(ASTNode::ContractDefinition(contract)) =
+                        symbol_place.closest_ancestor_of_type(context, NodeType::ContractDefinition)
+                    {
+                        graph.mark_used_pathways(source_unit.id, contract.id);
+                    }
+                }
+            }
+        }
+
+        for unused_import_id in graph.collect_unused_imports() {
+            if let Some(node) = context.nodes.get(&unused_import_id) {
+                capture!(self, context, node);
+            }
+        }
+        Ok(!self.found_instances.is_empty())
+    }
+
+    fn severity(&self) -> IssueSeverity {
+        IssueSeverity::Low
+    }
+
+    fn title(&self) -> String {
+        String::from("Unused Imports")
+    }
+
+    fn description(&self) -> String {
+        String::from("Redundant import statement. Consider removing it.")
+    }
+
+    fn instances(&self) -> BTreeMap<(String, usize, String), NodeID> {
+        self.found_instances.clone()
+    }
+
+    fn name(&self) -> String {
+        format!("{}", IssueDetectorNamePool::UnusedImport)
+    }
+}
+
+mod source_unit_graph_analysis {
+
+    //! Goal of this module is to create a graphical representation of all the source units
+    //! connected  with import statements.
+    //!
+    //! Will be used to detect unused imports
+
+    use std::collections::{HashMap, HashSet};
+
+    use crate::ast::NodeID;
+
+    #[derive(Default, Debug)]
+    pub struct GNode {
+        #[allow(dead_code)]
+        source_unit: NodeID,
+        edges: Vec<GEdge>,
+    }
+
+    #[derive(Default, Debug)]
+    pub struct GEdge {
+        symbols: Vec<NodeID>,
+        to: NodeID,
+        import_statement: NodeID,
+    }
+
+    #[derive(Default, Debug)]
+    pub struct Graph {
+        /// Key source units: source unit ID
+        source_units: HashMap<NodeID, GNode>,
+
+        /// Key: Import Statement, Value: Symbol that is imported
+        useful_symbols: HashMap<NodeID, Vec<NodeID>>,
+    }
+
+    impl Graph {
```

Comment:
Would this be useful to other detectors if we made it reusable like the preprocessed callgraph @TilakMaddy ?

Reply:
Possibly! But for now it does feel like a very specific graph for this detector only ...